apps/frontend/src/app/PageCharacter/CharacterDisplay/Tabs/TabTheorycraft/optimizeTc.ts

```diff
@@ -57,92 +78,213 @@ export function optimizeTc(
       if (f.operation === 'read' && f.path[0] === 'dyn') {
         const a = charTC.artifact.sets[f.path[1]]
         if (a) return constant(a)
-        if (!allSubstatKeys.includes(f.path[1] as any)) return constant(0)
+        if (!(allSubstatKeys as readonly string[]).includes(f.path[1]))
+          return constant(0)
       }
       return f
     },
     (f) => f
   )
   nodes = optimize(nodes, {}, (_) => false)
+  return {
+    nodes,
+  }
+}
+
+export function optimizeTcUsingNodes(
+  nodes: OptNode[],
+  charTC: ICharTC,
+  callback: (r: TCWorkerResult) => void
+) {
+  const startTime = performance.now()
+  const {
+    artifact: {
+      slots,
+      substats: { stats: substats, type: substatsType, rarity },
+    },
+    optimization: { distributedSubstats, maxSubstats, minTotal },
+  } = charTC
 
-  const subs = new Set<string>()
+  const scalesWith = new Set<string>()
   const compute = precompute(
     nodes,
     {},
     (f) => {
-      subs.add(f.path[1])
-      return f.path[1]
+      const val = f.path[1]
+      scalesWith.add(val)
+      return val
     },
     2
```

Comment:
Does it make sense to have 2? further down in perms, we feed the "existing subs" + "dynamic subs" as 2 separate "slots", but since "existing subs" is static, can we reduce this down to one, and feed the "existing subs" in the `initial` field?

Reply:
If some slots are static, combine them and feed to `initial`.